### PR TITLE
[DriveNet] Add close loop notebook

### PR DESCRIPTION
**Pull request by lucabergamini**
Add notebook for close loop evaluation of DriveNet

### CODE REVIEW

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 1]
```diff
@@ -0,0 +1,326 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# DriveNet Close Loop Evaluation\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:7]: Closed-loop

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "**Note: this notebook assumes you've already run the [training notebook](./drivenet_train.ipynb) and stored your model successfully.**\n",
+    "\n",
+    "## What is a close loop evaluation?\n",
+    "In close loop evaluation DriveNet is in **full control of the AV**. At each time step, we predict the future trajectory and then we move the AV int he first of the DriveNet's predictions. \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:12]: and then (remove we) move the ... to the first predicted position of DriveNet.

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "At Lyft, we refer to this process with the term **unroll**\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:14]: I think: 'by the term' ... are we keeping unroll? (I think in the paper we have 'forward-simulate')

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "\n",
+    "## What is a good close loop metric?\n",
+    "For this setting metrics are particularly challenging. In fact, we would like to penalise some of the drifting (e.g. going off road or in the opposite lane) while at the same time allow others (e.g. different speed profiles)\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:18]: simulation drift

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "At Lyft L5, we use a substantial sets of different metrics to capture dangerous manoeuvres and behaviours. \n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:20]: Maybe could extend this a bit longer, introduce some more motivation?
E.g. our real measurement would be put it on car ... but costly ... so try to mimick real world and so on - this is why we do closed-loop.

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "For the sake of simplicity, in this notebook we will be using a very simple proxy to detect if our model is driving in a sensible way, composed of two different metrics:\n",
+    "\n",
+    "### Collisions\n",
+    "Our AV should avoid collisions with other agents. This sounds trivial, but it's actually more complex than how it looks. In fact, while our AV is fully controlled by Drivenet, other agents are not (a setting we call **log replay**).\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:25]: than (remove how) it ...

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "If our AV was slower than the recorded one, chasing agents might bump into us. Clearly, this won't happen in a real setting where other agents can react to our behaviours.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:27]: behavior (singular)

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "Still, this metric is useful to assess how much our DriveNet follows the rules of the road.\n",
+    "\n",
+    "However, if we only considered collision, we might under penalise some potentially dangerous situations like driving off-road.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:31]: I don't like how this paragraph is structured: 'not as trivial as it sounds' -> this makes me more think of is detecting collisions trivial? Also, bad rear collisions are more caused by log-replay and not because this metric is bad ... ping me if unclear.

Comment (oliver-scheel) [reply examples/drivenet/drivenet_close_loop.ipynb:31]: Under penalised: it's not penalized at all then - would also reformulate this.

Comment (oliver-scheel) [reply examples/drivenet/drivenet_close_loop.ipynb:31]: collisionS

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    "\n",
+    "### Distance from Reference Trajectory\n",
+    "To address the issue presented above, we require our DriveNet to loosely stick to the original trajectory in the data. By setting the right threshold on the distance we can allow for different speed profile and small steerings, while pensalising large deviations like driving off-road.\n",
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:34]: different speed profileS,
would say: and small steering variations

Comment (oliver-scheel) [reply examples/drivenet/drivenet_close_loop.ipynb:34]: penalizing

File: examples/drivenet/drivenet_close_loop.ipynb [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    "model = model.eval()\n",
+    "torch.set_grad_enabled(False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load the Evaluation Data\n",
+    "Differently from training and open loop evaluation, this setting is intrinsically sequential. As such, we won't be using any parallelisation offered by pytorch."
```
Comment (oliver-scheel) [inline examples/drivenet/drivenet_close_loop.ipynb:118]: Different to

Comment (oliver-scheel) [reply examples/drivenet/drivenet_close_loop.ipynb:118]: Pytorch
